design.md

```diff
@@ -0,0 +1,106 @@
+
+## PoC
+
+https://github.com/wangxiaoxuan273/oras/pull/1
+
+Only implemented `oras manifest index create`.
+
+## Design
+### All commands
+`oras manifest index create` Create an image index from source manifests. 
+
+`oras manifest index update` Add/Remove manifests to/from an image index.
+
+### oras manifest index create
+
+Create an image index from source manifests. The command auto-detects platform information for each source manifests. The sources and the target index must be in the same repository.
+
+#### Usage
+
+`oras manifest index create [flags] <name>[:<tag[,<tag>][...]] [{<tag>|<digest>}...]`
+
+#### Flags
+
+(These flags are not implemented in the PoC)
+
+`--subject` Add a subject manifest for the to-be-created index.
+
+`--oci-layout` Set the given repository as an OCI layout.
+
+`--annotation` Add annotations for the to-be-created index.
+
+`--artifact-type` Add artifact type information for the to-be-created index.
+
+(May be added in the future) `--output` Output the updated manifest to a location.
+
+#### Aliases
+
+`pack`
+
+#### Examples
+
+Create an index from source manifests tagged amd64, darwin, armv7 in the repository `localhost:5000/hello`, and push the index without tagging it. We require that amd64, darwin, armv7 exist in `localhost:5000/hello` :
+`oras manifest index create localhost:5000/hello amd64 darwin armv7`
+
+Create an index from source manifests tagged amd64, darwin, armv7 in the repository localhost:5000/hello, and push the index with tag `latest`:
+`oras manifest index create localhost:5000/hello:latest amd64 darwin armv7`
+
+Create an index from source manifests using both tags and digests, and push the index with tag `latest`:
+`oras manifest index create localhost:5000/hello:latest amd64 sha256:xxx armv7`
+
+### oras manifest index update
+
+Add/Remove a manifest from an image index. The updated index will be created as a new index and the old index will not be deleted. 
+If the user specify the index with a tag, the corresponding tag will be updated to the new index. If the old index has other tags, the remaining tags will not be updated to the new index.
+
+#### Usage
+
+`oras manifest index update <name>{:<tag>|@<digest>} {--add/--remove} [{<tag>|<digest>}...]`
+
+#### Flags
+
+`--add` (shorthand `-a`) Add a manifest to the index. The manifest will be added as the last element of the index.
+
+`--remove` (shorthand `-r`) Remove a manifest from the index.
+
+**Note: One of the above flags should be used, as there has to be something to update. Otherwise the command does nothing.**
+
+`--oci-layout` Set the target as an oci image layout.
+
+`--tag` Tag the updated index. Multiple tags can be provided.
+
+(May be added in the future) `--output` Output the updated manifest to a location.
+
+#### Examples
+
+Add one manifest and remove two manifests from an index.
+
+`oras manifest index update localhost:5000/hello:latest --add win64 --remove sha256:xxx --remove arm64`
```

Comment:
It's interesting that we support referencing manifests by tags in the `index udpate` command🧐. In terms of implementation, does it mean we will first resolve the tag to get its digest, and then find the digest in the index?

Reply:
`GET /v2/<name>/manifests/<reference>` can be used to get the index without extra resolve operation.

Reply:
Oh I just realized its for `--add`. I was thinking of `--remove`, it looked strange to me to just remove a manifest by its tag.

Reply:
> It's interesting that we support referencing manifests by tags in the `index udpate` command🧐. In terms of implementation, does it mean we will first resolve the tag to get its digest, and then find the digest in the index?

Yes that's how I implemented it in my PoC. It is indeed strange to remove a manifest by its tag, but I think we should support it?